service/history/replication/task_hydrator.go

```diff
@@ -275,3 +283,47 @@ func (l mutableStateLoader) GetMutableState(ctx context.Context, domainID, workf
 func timeToUnixNano(t time.Time) *int64 {
 	return common.Int64Ptr(t.UnixNano())
 }
+
+type immediateHistoryProvider struct {
+	blob     *persistence.DataBlob
+	nextBlob *persistence.DataBlob
+}
+
+func (h immediateHistoryProvider) GetEventBlob(ctx context.Context, task persistence.ReplicationTaskInfo) (*types.DataBlob, error) {
+	if h.blob == nil {
+		return nil, errors.New("history blob not set")
+	}
+	return h.blob.ToInternal(), nil
+}
+
+func (h immediateHistoryProvider) GetNextRunEventBlob(ctx context.Context, task persistence.ReplicationTaskInfo) (*types.DataBlob, error) {
+	if h.nextBlob == nil {
+		return nil, nil // Expected and common
+	}
+	return h.nextBlob.ToInternal(), nil
+}
+
+type immediateMutableStateProvider struct {
+	ms immediateMutableState
+}
+
+func (r immediateMutableStateProvider) GetMutableState(ctx context.Context, domainID, workflowID, runID string) (mutableState, execution.ReleaseFunc, error) {
+	return r.ms, execution.NoopReleaseFn, nil
+}
+
+type immediateMutableState struct {
+	activities       map[int64]*persistence.ActivityInfo
+	versionHistories *persistence.VersionHistories
+}
+
+func (ms immediateMutableState) IsWorkflowExecutionRunning() bool {
+	// Immediate mutable state is always running
+	return true
```

Comment:
What is the impact if the workflow get closed immediately before replication?

Reply:
After some consideration I've update this.

This is used as early return when hydrating sync activity tasks. I think this check is somewhat redundant, since there will be no pending activity left after completion, and replication task will not be hydrated either way.

But for now let's keep it here for consistency with regular hydration from loaded mutable state. I may remove this entirely from both immediate and regular hydration later.